tower-ready-cache/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+                    debug_assert!(_cancel.is_some(), "missing cancelation");
+                    self.ready_services.insert(key, svc);
+                }
+                Err((key, UnreadyError::Canceled)) => {
+                    debug_assert!(!self.cancelations.contains_key(&key));
+                }
+                Err((key, UnreadyError::Inner(e))) => {
+                    debug!("dropping failed endpoint: {:?}", e);
+                    let _cancel = self.cancelations.swap_remove(&key);
+                    debug_assert!(_cancel.is_some());
+                }
+            }
+        }
+    }
+
+    /// Polls a service in the ready set to determine if it is still ready.
+    ///
+    /// If the service is no longer ready and there are no equivalent unready
+    /// services, it is moved back into the ready set and `Async::NotReady` is
+    /// returned.
+    ///
+    /// If the service errors, it is removed and dropped and the error is returned.
+    ///
+    /// Otherwise, `Async::Ready` is returned.
+    ///
+    /// Panics if the provided `index` is out of range.
+    pub fn poll_ready_index(&mut self, index: usize) -> Poll<(), Error> {
+        let (_, svc) = self
+            .ready_services
+            .get_index_mut(index)
+            .expect("index out of range");
+
+        match svc.poll_ready() {
+            Ok(Async::Ready(())) => Ok(Async::Ready(())),
+            Ok(Async::NotReady) => {
+                // became unready; so move it back there.
+                let (key, svc) = self
+                    .ready_services
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+                // If a new version of this service has been added to the
+                // unready set, don't overwrite it.
+                if !self.unready_contains(&key) {
+                    self.push_unready(key, svc);
+                }
+                Ok(Async::NotReady)
+            }
+            Err(e) => {
+                // failed, so drop it.
+                let e = e.into();
+                debug!("evicting failed endpoint: {:?}", e);
+                self.ready_services
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+                Err(e)
+            }
+        }
+    }
+
+    /// Swap-removes the indexed service from the ready set.
+    ///
+    /// Panics if the provided `index` is out of range.
+    pub fn process_ready_index<T, P>(&mut self, index: usize, process: P) -> T
+    where
+        P: FnOnce(&mut S) -> T,
+    {
+        let (key, mut svc) = self
+            .ready_services
+            .swap_remove_index(index)
+            .expect("index out of range");
+
+        let t = (process)(&mut svc);
+
+        // If a new version of this service has been added to the
+        // unready set, don't overwrite it.
+        if !self.unready_contains(&key) {
+            self.push_unready(key, svc);
```

Comment:
What if the service is immediately ready again even after `process`? Wouldn't it be a mistake to place it back in `unready`?

Reply:
The contract is that `poll_unready` must be called following this.

This is much like poll_ready vs call -- `call` can never invoke `poll` (as call need not be invoked from a task context).  process need not be called from a task context.